services/motion/explore/explore.go

```diff
@@ -0,0 +1,500 @@
+func NewExplore(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger) (motion.Service, error) {
+	ms := &explore{
+		Named:                 conf.ResourceName().AsNamed(),
+		logger:                logger,
+		obstacleResponseChan:  make(chan moveResponse),
+		executionResponseChan: make(chan moveResponse),
+		backgroundWorkers:     &sync.WaitGroup{},
+	}
+
+	if err := ms.Reconfigure(ctx, deps, conf); err != nil {
+		return nil, err
+	}
+	return ms, nil
+}
+
+// Reconfigure updates the explore motion service when the config has changed.
+func (ms *explore) Reconfigure(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+) (err error) {
+	ms.mutex.Lock()
+	defer ms.mutex.Unlock()
+
+	// Iterate over dependencies and store components and services along with the frame service directly
+	components := make(map[resource.Name]resource.Resource)
+	services := make(map[resource.Name]resource.Resource)
+	for name, dep := range deps {
+		switch dep := dep.(type) {
+		case framesystem.Service:
+			ms.fsService = dep
+		case vision.Service:
+			services[name] = dep
+		default:
+			components[name] = dep
+		}
+	}
+
+	ms.components = components
+	ms.services = services
+	return nil
+}
+
+type explore struct {
+	resource.Named
+	servicepb.UnimplementedMotionServiceServer
+
+	frameSystem referenceframe.FrameSystem
+	fsService   framesystem.Service
+	components  map[resource.Name]resource.Resource
+	services    map[resource.Name]resource.Resource
+	logger      logging.Logger
+	mutex       sync.Mutex
+
+	obstacleResponseChan  chan moveResponse
+	executionResponseChan chan moveResponse
+	backgroundWorkers     *sync.WaitGroup
+}
+
+func (ms *explore) Close(ctx context.Context) error {
+	utils.FlushChan(ms.obstacleResponseChan)
+	utils.FlushChan(ms.executionResponseChan)
+	ms.backgroundWorkers.Wait()
+	return nil
+}
+
+// Move takes a goal location and will plan and execute a movement to move a component specified by its name
+// to that destination.
+func (ms *explore) Move(
+	ctx context.Context,
+	componentName resource.Name,
+	destination *referenceframe.PoseInFrame,
+	worldState *referenceframe.WorldState,
+	constraints *servicepb.Constraints,
+	extra map[string]interface{},
+) (bool, error) {
+	ms.mutex.Lock()
+	defer ms.mutex.Unlock()
+
+	operation.CancelOtherWithLabel(ctx, exploreOpLabel)
+
+	// Parse extras
+	motionCfg, err := parseMotionConfig(extra)
+	if err != nil {
+		return false, err
+	}
+
+	// obstacleDetectors
+	obstacleDetectors, err := ms.createObstacleDetectors(motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create kinematic base
+	kb, err := ms.createKinematicBase(ctx, componentName, motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create motionplan plan
+	plan, err := ms.createMotionPlan(ctx, kb, destination, extra)
+	if err != nil {
+		return false, err
+	}
+
+	// Start background processes
+	cancelCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
cancel may need to be saved in explore struct in order for close to call it. Currently if the parent context is canceled (which it is in nav builtIn) this will close the child cancelCtx. However, if someone were to try to close explore directly, it would hang while waiting for background workers.

Reply:
Closing this resource on its own should not hang

Reply:
Tracking cancel function and adding to Close function